contrib/core-contract-tests/tests/sip-031/sip-031.fast-check.test.ts

```diff
@@ -0,0 +1,223 @@
+      accounts.wallet_4.address,
+    ),
+    accounts.wallet_5.address,
+  );
+  txOk(
+    indirectContract.transferStx(
+      constants.INITIAL_MINT_AMOUNT / 2n,
+      accounts.wallet_4.address,
+    ),
+    accounts.wallet_6.address,
+  );
+  // Mint the entire INITIAL_MINT_AMOUNT to the vesting contract
+  mint(constants.INITIAL_MINT_AMOUNT);
+}
+
+function months(n: number) {
+  return n * Number(constants.INITIAL_MINT_VESTING_ITERATION_BLOCKS);
+}
+
+test('property: vesting calculations are always mathematically correct', async () => {
+  await fc.assert(
+    fc.asyncProperty(
+      fc.integer({
+        min: 0,
+        max: 100 * Number(constants.INITIAL_MINT_VESTING_ITERATION_BLOCKS),
+      }), // blocks elapsed (0-100 months)
+      fc.bigInt({ min: 1n, max: 1000000n * 1000000n }), // extra deposit (1 micro-STX to 1M STX)
+      async (blocksElapsed, extraDeposit) => {
+        const manifestPath = global.options.clarinet.manifestPath;
+        await simnet.initSession(process.cwd(), manifestPath);
+
+        const monthsElapsed = Math.floor(
+          blocksElapsed /
+            Number(constants.INITIAL_MINT_VESTING_ITERATION_BLOCKS),
+        );
+
+        console.log(
+          `Running vesting calculation test for ${blocksElapsed} blocks and extra deposit ${extraDeposit}`,
+        );
+        mintInitial();
+
+        // Add extra deposit
+        if (extraDeposit > 0n) {
+          mint(extraDeposit);
+        }
+
+        // Advance time
+        if (monthsElapsed > 0) {
+          simnet.mineEmptyBlocks(months(monthsElapsed));
+        }
+
+        // Calculate expected vested amount
+        const effectiveMonths = Math.min(monthsElapsed, 24);
+        const expectedVested =
+          effectiveMonths < 24
+            ? (constants.INITIAL_MINT_VESTING_AMOUNT /
+                constants.INITIAL_MINT_VESTING_ITERATIONS) *
+              BigInt(effectiveMonths)
+            : constants.INITIAL_MINT_VESTING_AMOUNT;
+
+        const expectedTotal =
+          constants.INITIAL_MINT_IMMEDIATE_AMOUNT +
+          expectedVested +
+          extraDeposit;
+
+        // Claim and verify
+        const receipt = txOk(contract.claim(), accounts.deployer.address);
+
+        // Properties that must always hold:
+        // 1. Claimed amount should match calculation
+        expect(receipt.value).toBe(expectedTotal);
+
+        // 2. Remaining balance should be correct
+        const remainingBalance = rov(
+          indirectContract.getBalance(contract.identifier),
+        );
+        const expectedRemaining =
+          effectiveMonths < 24
+            ? constants.INITIAL_MINT_VESTING_AMOUNT - expectedVested
+            : 0n;
+        expect(remainingBalance).toBe(expectedRemaining);
+
+        // 3. Total funds should be conserved
+        const totalFunds = receipt.value + remainingBalance;
+        expect(totalFunds).toBe(constants.INITIAL_MINT_AMOUNT + extraDeposit);
+      },
+    ),
+    { numRuns: 50 },
```

Comment:
Each test overrides the default (100) `numRuns`, probably to speed up CI(?). Better to unify with an env var:

```ts
numRuns: process.env.FC_NUM_RUNS
  ? Number(process.env.FC_NUM_RUNS)
  : undefined
```

CI can set `FC_NUM_RUNS=20`. Falls back to fast-check default when unset. There's also a global config for this if needed:

```ts
fc.configureGlobal({
  numRuns: process.env.FC_NUM_RUNS
    ? Number(process.env.FC_NUM_RUNS)
    : undefined
});
```

Reply:
Sure, we can do that if it's simpler. I had manually set run numbers for these to get each to run for a reasonable amount of time.